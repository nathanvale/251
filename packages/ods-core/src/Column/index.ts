export { Column } from "./Column";
export type ColumnProps = import("./Column").ColumnProps;
