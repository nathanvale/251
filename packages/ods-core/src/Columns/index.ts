export { Columns, ColumnsContext } from "./Columns";
export type ColumnsProps = import("./Columns").ColumnsProps;
export type CollapsibleAlignmentChildProps = import("./Columns").CollapsibleAlignmentChildProps;
