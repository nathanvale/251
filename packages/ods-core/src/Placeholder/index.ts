export { Placeholder } from "./Placeholder";
export type PlaceholderProps = import("./Placeholder").PlaceholderProps;
