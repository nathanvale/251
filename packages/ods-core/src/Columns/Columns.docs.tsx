/* eslint-disable react/display-name */
import React from 'react';
import {ComponentDocs} from '../../../../docs/src/types';
import {Placeholder, Columns, Column, ColumnsProps} from '..';

export const docs: ComponentDocs<ColumnsProps> = {
  category: 'Layout',
  componentName: 'Columns',
  description:
    'Columns is a low-level atomic layout component. Its main concerns are laying out columns in a 12 grid layout and responsively stacking them at a set breakpoint. Open the examples in Playroom to see the responsive behaviour of this component.',
  propDescriptions: {},
  migrationGuide: false,
  examples: [
    {
      label: 'Responsively Collapsing Columns to a Stack',
      description:
        "Columns can collapse from a horizontal layout to a vertical stack by passing in a 'collapseBelow' breakpoint. In this example all columns below the 'sm' breakpoint will collapse to a stack. Open in Playroom to see the responsive behaviour of this example.",
      Code: () => (
        <Columns space="small" collapseBelow="sm">
          <Column>
            <Placeholder backgroundColor="blue" />
          </Column>
          <Column>
            <Placeholder backgroundColor="blue" />
          </Column>
        </Columns>
      ),
    },
    {
      label: 'Responsively Collapsing Columns Below Desktop',
      description:
        'In this exmaple columns collapse to a stack below the desktop breakpoint. Open in Playroom to see the responsive behaviour of this example.',
      Code: () => (
        <Columns space="small" collapseBelow="lg">
          <Column>
            <Placeholder backgroundColor="blue" />
          </Column>
          <Column>
            <Placeholder backgroundColor="blue" />
          </Column>
        </Columns>
      ),
    },
    {
      label: 'Responsive Space on Mobile and Desktop Breakpoints',
      description:
        'Distributed space can be controlled responsively by passing in an array of exactly two tshirt sizes - one for mobile and one for desktop. Open in Playroom to see the responsive behaviour of this example.',
      Code: () => (
        <Columns space={['small', 'xxxlarge']}>
          <Column>
            <Placeholder backgroundColor="blue" />
          </Column>
          <Column>
            <Placeholder backgroundColor="blue" />
          </Column>
        </Columns>
      ),
    },
    {
      label: 'Responsive Space Across All Five Breakpoints',
      description:
        'For more granular control across all 5 breakpoints, xs, sm, md, lg and xl, distributed space can be controlled responsively by passing in an object with breakpoints as keys and tshirt sizes as values. Open in Playroom to see the responsive behaviour of this example.',
      Code: () => (
        <Columns
          space={{
            xs: 'none',
            sm: 'small',
            md: 'medium',
            lg: 'large',
            xl: 'xxlarge',
          }}
        >
          <Column>
            <Placeholder backgroundColor="blue" />
          </Column>
          <Column>
            <Placeholder backgroundColor="blue" />
          </Column>
        </Columns>
      ),
    },
    {
      label: 'Vertical Alignment',
      description:
        'Use alignY to align the child Column components top, center (defalut) or bottom',
      Code: () => (
        <Columns space="small" collapseBelow="sm" alignY="bottom">
          <Column>
            <Placeholder backgroundColor="blue" />
          </Column>
          <Column>
            <Placeholder backgroundColor="blue" />
          </Column>
        </Columns>
      ),
    },
  ],
};
