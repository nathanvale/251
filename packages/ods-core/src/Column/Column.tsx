/* eslint-disable no-eval */
import React, { ReactNode, useContext } from "react";
import styled, { css } from "styled-components";
import {
  mapSpaceAliasToIndex,
  setBreakpoint,
} from "@origin-digital/ods-helpers";
import { ColumnsContext } from "../Columns/Columns";
import { Box } from "../Box/Box";
import { BoxDebug } from "../_private/components/BoxDebug/BoxDebug";

export interface ColumnProps {
  children: ReactNode;
  width?:
    | "content"
    | "1/2"
    | "1/3"
    | "2/3"
    | "1/4"
    | "3/4"
    | "1/6"
    | "5/6"
    | "1/12"
    | "5/12"
    | "7/12"
    | "11/12"
    | "flex";
  "data-id"?: string;
}

const OuterStyledBox = styled(Box)<{ columnWidth: ColumnProps["width"] }>`
  min-width: 0;
  ${p => (p.columnWidth === "content" ? "flex-shrink: 0;" : undefined)}
  ${p =>
    p.columnWidth && p.columnWidth !== "content" && p.columnWidth !== "flex"
      ? css`
          flex: 0 0 ${eval(p.columnWidth) * 100}%;
        `
      : undefined}
`;

const InnerStyledBox = styled(Box)`
  div:first-child > & {
    padding-top: 0;
  }
`;

export const Column = ({
  children,
  width = "flex",
  "data-id": dataId,
}: ColumnProps) => {
  const { collapseBelow, space } = useContext(ColumnsContext);
  const spaceIndex = space && mapSpaceAliasToIndex({ space });
  return (
    <OuterStyledBox
      data-id={dataId}
      columnWidth={width}
      width={width !== "content" ? "full" : undefined}
    >
      <InnerStyledBox
        height="full"
        paddingLeft={
          (collapseBelow
            ? setBreakpoint({
                breakpoint: collapseBelow,
                value1: "none",
                value2: spaceIndex,
              })
            : spaceIndex) as TS_FIXME
        }
        paddingTop={
          (collapseBelow
            ? setBreakpoint({
                breakpoint: collapseBelow,
                value1: spaceIndex,
                value2: "none",
              })
            : "none") as TS_FIXME
        }
      >
        <BoxDebug>{children}</BoxDebug>
      </InnerStyledBox>
    </OuterStyledBox>
  );
};

Column.defaultProps = {
  "data-id": "column",
  width: "flex",
};

Column.displayName = "Column";
