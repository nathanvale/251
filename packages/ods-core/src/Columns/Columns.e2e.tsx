import * as React from "react";
import { E2ETests } from "@origin-digital/ods-types";
import { Column } from "../Column/Column";
import { Placeholder } from "../Placeholder/Placeholder";
import { Columns } from "./Columns";

export const tests: E2ETests = [
  {
    label: "AlignY Bottom",
    Code: () => (
      <Columns alignY="bottom">
        <Column>
          <Placeholder height="100px" />
        </Column>
        <Column>
          <Placeholder height="40px" />
        </Column>
        <Column>
          <Placeholder height="70px" />
        </Column>
      </Columns>
    ),
  },
  {
    label: "AlignY Center",
    Code: () => (
      <Columns alignY="center">
        <Column>
          <Placeholder height="100px" />
        </Column>
        <Column>
          <Placeholder height="40px" />
        </Column>
        <Column>
          <Placeholder height="70px" />
        </Column>
      </Columns>
    ),
  },
  {
    label: "Default AlignY - Top",
    Code: () => (
      <Columns>
        <Column>
          <Placeholder height="100px" />
        </Column>
        <Column>
          <Placeholder height="40px" />
        </Column>
        <Column>
          <Placeholder height="70px" />
        </Column>
      </Columns>
    ),
  },
  {
    label: "Space xxsmall",
    Code: () => (
      <Columns space="xxsmall">
        <Column>
          <Placeholder height="100px" />
        </Column>
        <Column>
          <Placeholder height="40px" />
        </Column>
        <Column>
          <Placeholder height="70px" />
        </Column>
      </Columns>
    ),
  },
  {
    label: "Space xsmall",
    Code: () => (
      <Columns space="xsmall">
        <Column>
          <Placeholder height="100px" />
        </Column>
        <Column>
          <Placeholder height="40px" />
        </Column>
        <Column>
          <Placeholder height="70px" />
        </Column>
      </Columns>
    ),
  },
  {
    label: "Space small",
    Code: () => (
      <Columns space="small">
        <Column>
          <Placeholder height="100px" />
        </Column>
        <Column>
          <Placeholder height="40px" />
        </Column>
        <Column>
          <Placeholder height="70px" />
        </Column>
      </Columns>
    ),
  },
  {
    label: "Space medium",
    Code: () => (
      <Columns space="medium">
        <Column>
          <Placeholder height="100px" />
        </Column>
        <Column>
          <Placeholder height="40px" />
        </Column>
        <Column>
          <Placeholder height="70px" />
        </Column>
      </Columns>
    ),
  },
  {
    label: "Space large",
    Code: () => (
      <Columns space="large">
        <Column>
          <Placeholder height="100px" />
        </Column>
        <Column>
          <Placeholder height="40px" />
        </Column>
        <Column>
          <Placeholder height="70px" />
        </Column>
      </Columns>
    ),
  },
  {
    label: "Space xlarge",
    Code: () => (
      <Columns space="xlarge">
        <Column>
          <Placeholder height="100px" />
        </Column>
        <Column>
          <Placeholder height="40px" />
        </Column>
        <Column>
          <Placeholder height="70px" />
        </Column>
      </Columns>
    ),
  },
  {
    label: "Space xxlarge",
    Code: () => (
      <Columns space="xxlarge">
        <Column>
          <Placeholder height="100px" />
        </Column>
        <Column>
          <Placeholder height="40px" />
        </Column>
        <Column>
          <Placeholder height="70px" />
        </Column>
      </Columns>
    ),
  },
  {
    label: "Space xxxlarge",
    Code: () => (
      <Columns space="xxxlarge">
        <Column>
          <Placeholder height="100px" />
        </Column>
        <Column>
          <Placeholder height="40px" />
        </Column>
        <Column>
          <Placeholder height="70px" />
        </Column>
      </Columns>
    ),
  },
  {
    label: "Responsive space",
    responsive: true,
    Code: () => (
      <Columns space={["xsmall", "xlarge"]}>
        <Column>
          <Placeholder height="100px" />
        </Column>
        <Column>
          <Placeholder height="40px" />
        </Column>
        <Column>
          <Placeholder height="70px" />
        </Column>
      </Columns>
    ),
  },
  {
    label: "Collapse bellow md",
    responsive: true,
    Code: () => (
      <Columns collapseBelow="md" space="medium">
        <Column>
          <Placeholder height="100px" />
        </Column>
        <Column>
          <Placeholder height="40px" />
        </Column>
        <Column>
          <Placeholder height="70px" />
        </Column>
      </Columns>
    ),
  },
  {
    label: "Nested Columns",
    responsive: true,
    Code: () => (
      <>
        <Columns space="none">
          <Column>
            <Columns space="none">
              <Column>
                <Placeholder width="100%" height="160px" />
              </Column>
              <Column>
                <Placeholder minWidth="36px" height="160px" />
              </Column>
            </Columns>
          </Column>
          <Column>
            <Columns space="none">
              <Column>
                <Placeholder minWidth="36px" height="160px" />
              </Column>
              <Column>
                <Placeholder minWidth="36px" height="160px" />
              </Column>
            </Columns>
          </Column>
        </Columns>
        <Columns space="none">
          <Column>
            <Columns space="none">
              <Column>
                <Placeholder minWidth="36px" height="160px" />
              </Column>
              <Column>
                <Placeholder minWidth="36px" height="160px" />
              </Column>
            </Columns>
          </Column>
          <Column>
            <Columns space="none">
              <Column>
                <Placeholder minWidth="36px" height="160px" />
              </Column>
              <Column>
                <Placeholder minWidth="36px" height="160px" />
              </Column>
            </Columns>
          </Column>
        </Columns>
      </>
    ),
  },
];
