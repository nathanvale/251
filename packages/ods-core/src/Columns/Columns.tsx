import React, { ReactElement, createContext } from "react";
import {
  ResponsiveSpace,
  CollapsibleAlignmentProps,
  OptionalTrackableProps,
  SpaceVariants,
} from "@origin-digital/ods-types";
import {
  normaliseResponsiveProp,
  mapToStyledSystem,
} from "@origin-digital/ods-helpers";
import { ColumnProps } from "../Column/Column";

import { resolveCollapsibleAlignmentProps } from "../_private/resolveCollapsibleAlignmentProps";
import { InnerContainer } from "../Inline/Inline";

export type CollapsibleAlignmentChildProps = ReturnType<
  typeof resolveCollapsibleAlignmentProps
>["collapsibleAlignmentChildProps"];

interface ColumnsContextValue {
  collapseXs: boolean;
  collapseSm: boolean;
  collapseMd: boolean;
  collapseLg: boolean;
  collapseXl: boolean;
  xsSpace?: SpaceVariants;
  smSpace?: SpaceVariants;
  mdSpace?: SpaceVariants;
  lgSpace?: SpaceVariants;
  xlSpace?: SpaceVariants;
  /**
   * The display, flexDirection and justifyContent css props per xs,sm,md,lg or
   * xl breakpoint depending on if the Columns are collapsed or aligned or not.
   * This logic is shared with the Inline component.
   */
  collapsibleAlignmentChildProps: CollapsibleAlignmentChildProps | {};
}

export const ColumnsContext = createContext<ColumnsContextValue>({
  collapseXs: false,
  collapseSm: false,
  collapseMd: false,
  collapseLg: false,
  collapseXl: false,
  xsSpace: "none",
  smSpace: "none",
  mdSpace: "none",
  lgSpace: "none",
  xlSpace: "none",
  collapsibleAlignmentChildProps: {},
});

export interface ColumnsProps
  extends CollapsibleAlignmentProps,
    OptionalTrackableProps {
  space: ResponsiveSpace;
  children:
    | Array<ReactElement<ColumnProps> | null>
    | ReactElement<ColumnProps>
    | null;
}

const defaultSpace = "none";
const defaultAlignX = "left";
const defaultAlignY = "top";

export const Columns = ({
  children,
  collapseBelow,
  "data-id": dataId,
  space = defaultSpace,
  alignX = defaultAlignX,
  alignY = defaultAlignY,
}: ColumnsProps) => {
  const [xsSpace, smSpace, mdSpace, lgSpace, xlSpace] = normaliseResponsiveProp(
    space
  );

  const {
    collapsibleAlignmentProps,
    collapsibleAlignmentChildProps,
    collapseXs,
    collapseSm,
    collapseMd,
    collapseLg,
    collapseXl,
  } = resolveCollapsibleAlignmentProps({
    collapseBelow,
    alignX,
    alignY,
  });

  return (
    <InnerContainer
      data-id={dataId}
      space={
        mapToStyledSystem({
          xs: collapseXs ? "none" : xsSpace,
          sm: collapseSm ? "none" : smSpace,
          md: collapseMd ? "none" : mdSpace,
          lg: collapseLg ? "none" : lgSpace,
          xl: collapseXl ? "none" : xlSpace,
        }) || "none"
      }
      {...collapsibleAlignmentProps}
    >
      <ColumnsContext.Provider
        value={{
          collapseXs,
          collapseSm,
          collapseMd,
          collapseLg,
          collapseXl,
          xsSpace,
          smSpace,
          mdSpace,
          lgSpace,
          xlSpace,
          collapsibleAlignmentChildProps,
        }}
      >
        {children}
      </ColumnsContext.Provider>
    </InnerContainer>
  );
};

Columns.defaultProps = {
  "data-id": "columns",
  space: "large",
  alignY: defaultAlignY,
  alignX: defaultAlignX,
};

Columns.displayName = "Columns";
