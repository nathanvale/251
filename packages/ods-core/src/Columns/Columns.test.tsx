import React from "react";
import {
  queryAllByAttribute,
  render,
} from "@origin-digital/ods-testing-library";
import { Column } from "../Column";
import { Placeholder } from "../Placeholder";
import { Columns } from "./Columns";

test("it spaces column comps with small space", () => {
  const { container } = render(
    <Columns space="small">
      <Column>
        <Placeholder />
      </Column>
      <Column>
        <Placeholder />
      </Column>
      <Column>
        <Placeholder />
      </Column>
    </Columns>
  );
  //Parent container sets initial margin to negative of the space value.
  const style = window.getComputedStyle(container.firstChild as Element);
  expect(style.marginLeft).toEqual("-12px");

  // Each Column then sets a left padding of space (small = 12px).
  queryAllByAttribute("data-id", container, "column").forEach((colEl) => {
    const elStyle = window.getComputedStyle(colEl.firstChild as Element);
    expect(elStyle.paddingLeft).toEqual("12px");
  });
});

test("it top-aligns children by default", () => {
  const { container } = render(
    <Columns data-id="my-cols">
      <Column>
        <Placeholder height="100px" />
      </Column>
      <Column>
        <Placeholder height="40px" />
      </Column>
      <Column>
        <Placeholder height="70px" />
      </Column>
    </Columns>
  );

  expect(container).toMatchSnapshot();
});

test("it vertically aligns children middle if alignY is center", () => {
  const { container } = render(
    <Columns alignY="center">
      <Column>
        <Placeholder height="100px" />
      </Column>
      <Column>
        <Placeholder height="40px" />
      </Column>
      <Column>
        <Placeholder height="70px" />
      </Column>
    </Columns>
  );

  expect(container).toMatchSnapshot();
});

test("it vertically aligns children bottom if alignY is bottom", () => {
  const { container } = render(
    <Columns alignY="bottom">
      <Column>
        <Placeholder height="100px" />
      </Column>
      <Column>
        <Placeholder height="40px" />
      </Column>
      <Column>
        <Placeholder height="70px" />
      </Column>
    </Columns>
  );

  expect(container).toMatchSnapshot();
});
