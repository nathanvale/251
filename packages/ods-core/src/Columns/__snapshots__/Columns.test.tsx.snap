// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it top-aligns children by default 1`] = `
.c2 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  width: 100%;
}

.c5 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
}

.c7 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  background-color: #e3e3e3;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  overflow: hidden;
}

.c0 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: 0;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c4 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  padding-top: 0;
  padding-left: 0;
  height: 100%;
}

.c1 {
  min-width: 0;
}

div:first-child > .c3 {
  padding-top: 0;
}

.c6 {
  width: 100%;
  min-width: 40px;
  height: 100px;
  border: 2px solid rgba(51,51,51,0.3);
}

.c8 {
  width: 100%;
  min-width: 40px;
  height: 40px;
  border: 2px solid rgba(51,51,51,0.3);
}

.c9 {
  width: 100%;
  min-width: 40px;
  height: 70px;
  border: 2px solid rgba(51,51,51,0.3);
}

<div>
  <div
    class="c0"
    data-id="my-cols"
    display="flex"
  >
    <div
      class="c1 c2"
      data-id="column"
      width="full"
    >
      <div
        class="c3 c4"
        height="full"
      >
        <div
          class="c5"
        >
          <div
            class="c6 c7"
            data-id="placeholder"
            display="flex"
            height="100px"
            overflow="hidden"
            shape="rectangle"
            width="100%"
          >
            <svg
              style="width: 100%; height: 100%; position: absolute;"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="0"
                x2="100%"
                y1="0"
                y2="100%"
              />
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="100%"
                x2="0"
                y1="0"
                y2="100%"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c1 c2"
      data-id="column"
      width="full"
    >
      <div
        class="c3 c4"
        height="full"
      >
        <div
          class="c5"
        >
          <div
            class="c8 c7"
            data-id="placeholder"
            display="flex"
            height="40px"
            overflow="hidden"
            shape="rectangle"
            width="100%"
          >
            <svg
              style="width: 100%; height: 100%; position: absolute;"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="0"
                x2="100%"
                y1="0"
                y2="100%"
              />
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="100%"
                x2="0"
                y1="0"
                y2="100%"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c1 c2"
      data-id="column"
      width="full"
    >
      <div
        class="c3 c4"
        height="full"
      >
        <div
          class="c5"
        >
          <div
            class="c9 c7"
            data-id="placeholder"
            display="flex"
            height="70px"
            overflow="hidden"
            shape="rectangle"
            width="100%"
          >
            <svg
              style="width: 100%; height: 100%; position: absolute;"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="0"
                x2="100%"
                y1="0"
                y2="100%"
              />
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="100%"
                x2="0"
                y1="0"
                y2="100%"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`it vertically aligns children bottom if alignY is bottom 1`] = `
.c2 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  width: 100%;
}

.c5 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
}

.c7 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  background-color: #e3e3e3;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  overflow: hidden;
}

.c4 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  padding-top: 0;
  padding-left: 0;
  height: 100%;
}

.c0 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: 0;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c1 {
  min-width: 0;
}

div:first-child > .c3 {
  padding-top: 0;
}

.c6 {
  width: 100%;
  min-width: 40px;
  height: 100px;
  border: 2px solid rgba(51,51,51,0.3);
}

.c8 {
  width: 100%;
  min-width: 40px;
  height: 40px;
  border: 2px solid rgba(51,51,51,0.3);
}

.c9 {
  width: 100%;
  min-width: 40px;
  height: 70px;
  border: 2px solid rgba(51,51,51,0.3);
}

<div>
  <div
    class="c0"
    data-id="columns"
    display="flex"
  >
    <div
      class="c1 c2"
      data-id="column"
      width="full"
    >
      <div
        class="c3 c4"
        height="full"
      >
        <div
          class="c5"
        >
          <div
            class="c6 c7"
            data-id="placeholder"
            display="flex"
            height="100px"
            overflow="hidden"
            shape="rectangle"
            width="100%"
          >
            <svg
              style="width: 100%; height: 100%; position: absolute;"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="0"
                x2="100%"
                y1="0"
                y2="100%"
              />
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="100%"
                x2="0"
                y1="0"
                y2="100%"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c1 c2"
      data-id="column"
      width="full"
    >
      <div
        class="c3 c4"
        height="full"
      >
        <div
          class="c5"
        >
          <div
            class="c8 c7"
            data-id="placeholder"
            display="flex"
            height="40px"
            overflow="hidden"
            shape="rectangle"
            width="100%"
          >
            <svg
              style="width: 100%; height: 100%; position: absolute;"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="0"
                x2="100%"
                y1="0"
                y2="100%"
              />
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="100%"
                x2="0"
                y1="0"
                y2="100%"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c1 c2"
      data-id="column"
      width="full"
    >
      <div
        class="c3 c4"
        height="full"
      >
        <div
          class="c5"
        >
          <div
            class="c9 c7"
            data-id="placeholder"
            display="flex"
            height="70px"
            overflow="hidden"
            shape="rectangle"
            width="100%"
          >
            <svg
              style="width: 100%; height: 100%; position: absolute;"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="0"
                x2="100%"
                y1="0"
                y2="100%"
              />
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="100%"
                x2="0"
                y1="0"
                y2="100%"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`it vertically aligns children middle if alignY is center 1`] = `
.c2 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  width: 100%;
}

.c5 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
}

.c7 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  background-color: #e3e3e3;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  overflow: hidden;
}

.c4 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  padding-top: 0;
  padding-left: 0;
  height: 100%;
}

.c0 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  min-width: 0;
}

div:first-child > .c3 {
  padding-top: 0;
}

.c6 {
  width: 100%;
  min-width: 40px;
  height: 100px;
  border: 2px solid rgba(51,51,51,0.3);
}

.c8 {
  width: 100%;
  min-width: 40px;
  height: 40px;
  border: 2px solid rgba(51,51,51,0.3);
}

.c9 {
  width: 100%;
  min-width: 40px;
  height: 70px;
  border: 2px solid rgba(51,51,51,0.3);
}

<div>
  <div
    class="c0"
    data-id="columns"
    display="flex"
  >
    <div
      class="c1 c2"
      data-id="column"
      width="full"
    >
      <div
        class="c3 c4"
        height="full"
      >
        <div
          class="c5"
        >
          <div
            class="c6 c7"
            data-id="placeholder"
            display="flex"
            height="100px"
            overflow="hidden"
            shape="rectangle"
            width="100%"
          >
            <svg
              style="width: 100%; height: 100%; position: absolute;"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="0"
                x2="100%"
                y1="0"
                y2="100%"
              />
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="100%"
                x2="0"
                y1="0"
                y2="100%"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c1 c2"
      data-id="column"
      width="full"
    >
      <div
        class="c3 c4"
        height="full"
      >
        <div
          class="c5"
        >
          <div
            class="c8 c7"
            data-id="placeholder"
            display="flex"
            height="40px"
            overflow="hidden"
            shape="rectangle"
            width="100%"
          >
            <svg
              style="width: 100%; height: 100%; position: absolute;"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="0"
                x2="100%"
                y1="0"
                y2="100%"
              />
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="100%"
                x2="0"
                y1="0"
                y2="100%"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c1 c2"
      data-id="column"
      width="full"
    >
      <div
        class="c3 c4"
        height="full"
      >
        <div
          class="c5"
        >
          <div
            class="c9 c7"
            data-id="placeholder"
            display="flex"
            height="70px"
            overflow="hidden"
            shape="rectangle"
            width="100%"
          >
            <svg
              style="width: 100%; height: 100%; position: absolute;"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="0"
                x2="100%"
                y1="0"
                y2="100%"
              />
              <line
                style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
                x1="100%"
                x2="0"
                y1="0"
                y2="100%"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
